src/main/java/org/tron/core/services/WitnessProductBlockService.java

```diff
@@ -0,0 +1,96 @@
+package org.tron.core.services;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+import org.tron.common.utils.ByteArray;
+import org.tron.core.capsule.BlockCapsule;
+
+@Slf4j
+@Service
+public class WitnessProductBlockService {
+
+  private Cache<Long, BlockCapsule> historyBlockCapsuleCache = CacheBuilder.newBuilder()
+      .initialCapacity(200).maximumSize(200).build();
+
+  private Map<String, CheatWitnessInfo> cheatWitnessInfoMap = new HashMap<>();
+
+  public static class CheatWitnessInfo {
+
+    private AtomicInteger times = new AtomicInteger(0);
+    private long latestBlockNum;
+    private Set<BlockCapsule> blockCapsuleSet = new HashSet<>();
+
+    public AtomicInteger getTimes() {
+      return times;
+    }
+
+    public CheatWitnessInfo setTimes(AtomicInteger times) {
+      this.times = times;
+      return this;
+    }
+
+    public long getLatestBlockNum() {
+      return latestBlockNum;
+    }
+
+    public CheatWitnessInfo setLatestBlockNum(long latestBlockNum) {
+      this.latestBlockNum = latestBlockNum;
+      return this;
+    }
+
+    public Set<BlockCapsule> getBlockCapsuleSet() {
+      return blockCapsuleSet;
+    }
+
+    public CheatWitnessInfo setBlockCapsuleSet(
+        Set<BlockCapsule> blockCapsuleSet) {
+      this.blockCapsuleSet = blockCapsuleSet;
```

Comment:
![Vulnerability](https://sonarsource.github.io/sonar-github/vulnerability.png 'Vulnerability') Vulnerability: Store a copy of "blockCapsuleSet". ([squid:S2384](https://sonarcloud.io/organizations/tron-zhaohong/rules#rule_key=squid%3AS2384))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=java-tron&pullRequest=1639&issues=AWalNEdtdPPRVdwyi0o9&open=AWalNEdtdPPRVdwyi0o9)

[ISSUE_KEY]: AWalNEdtdPPRVdwyi0o9